src/charm.py

```diff
@@ -356,6 +356,31 @@ def _block_on_openstack_config(self, state: CharmState) -> bool:
             return True
         return False
 
+    def _common_install_code(self, state: CharmState) -> bool:
+        """Installation code shared between install and upgrade hook.
+
+        Returns:
+            Whether the installation code was successful.
+        """
+        self.unit.status = MaintenanceStatus("Installing packages")
+        try:
+            # The `_start_services`, `_install_deps` includes retry.
+            self._install_deps()
+            self._start_services(state.charm_config.token, state.proxy_config)
+            metrics.setup_logrotate()
+        except (LogrotateSetupError, SubprocessError) as err:
```

Comment:
I'm wondering if it would help to keep the exception scopes limited to the function that can raise it - i.e. try except block only for metrics.setup_logrotate for LogrotateSetupError and SubprocessError for apt-installs.